keps/sig-cluster-lifecycle/wgs/20200108-instance-specific-componentconfig.md

```diff
@@ -0,0 +1,418 @@
+component. The naming convention for the API group is `{component}.config.k8s.io`,
+and the convention for the Kind is `{Component}Configuration`. This KEP proposes
+that the naming convention for the new Kind be `{Component}InstanceConfiguration`.
+
+The canonical flag for providing a ComponentConfig file to a component is
+`--config`. This KEP proposes that the canonical flag for providing an
+instance-specific config file be `--instance-config`, and that the
+instance-specific object not initially be permitted as part of a yaml stream
+in the `--config` file (and vice-versa). This is for the sake of a simple
+implementation and can be enabled in the future, if we decide it is useful.
+
+As with sharable ComponentConfig parameters, command line flags for
+instance-specific parameters should continue to function and take precedence
+over values from the config file so that backwards compatibility is maintained.
+
+### Kube-proxy Example
+
+Today, a kube-proxy DaemonSet may be configued as follows to mix sharable and
+instance-specific configuration parameters (using flags only). Note some fields
+and arguments that would be set in production have been removed or modified
+below to improve the clarity of the example.
+
+```yaml
+apiVersion: apps/v1
+kind: DaemonSet
+  name: kube-proxy
+  namespace: kube-system
+spec:
+  template:
+    spec:
+      containers:
+      - name: kube-proxy
+        image: kube-proxy-amd64
+        command:
+        - kube-proxy
+        - --proxy-mode=ipvs
+        - --hostname-override=${NODE_NAME}
+        env:
+          - NODE_NAME
+            valueFrom:
+              fieldRef:
+                fieldPath: spec.nodeName
+```
+
+Below, how the above may be modified when moving to a shared ComponentConfig
+ConfigMap. Note that the instance-specific parameter is still specified via
+a flag, because the ConfigMap is not appropriate for instance-specific
+parameters.
+
+```yaml
+apiVersion: apps/v1
+kind: DaemonSet
+  name: kube-proxy
+  namespace: kube-system
+spec:
+  template:
+    spec:
+      containers:
+      - name: kube-proxy
+        image: kube-proxy-amd64
+        command:
+        - kube-proxy
+        - --config=/etc/config/kube-proxy.yaml
+        - --hostname-override=${NODE_NAME}
+        env:
+          - NODE_NAME
+            valueFrom:
+              fieldRef:
+                fieldPath: spec.nodeName
+        volumeMounts:
+        - mountPath: /etc/config
+          name: config
+          readOnly: true
+      volumes:
+      - name: config
+        configMap:
+          name: kube-proxy-config-ghkm248tbc
+```
+
+Finally, when instance-specific configuration is also provided via the
+ComponentConfig approach, using the separate file proposed above.
+
+```yaml
+apiVersion: apps/v1
+kind: DaemonSet
+  name: kube-proxy
+  namespace: kube-system
+spec:
+  template:
+    spec:
+      initContainers:
+      - name: instance-config
+        image: something-with-bash
```

Comment:
of topic:
with most / all of our images moving to distroless one might have to trust a third party image source to perform the instance specific amend. wondering if we should have a recommended image for this at the k8s GCR.

Reply:
I like the idea of having a recommended image. As you mention below, the process of manually creating the instance specific config is quite verbose and could be unintuitive for users ("Why do I need to create this file when I can just pass a flag?"). I could imagine providing a specific image that can be used as an init container which you just pass the values through environment variables would help ease the transition and increase adoption.

```yaml
 initContainers:
      - name: instance-config
        image: k8s.gcr.io/init-kube-proxy-instance-config:v1alpha1
        env:
          - NODE_NAME
            valueFrom:
              fieldRef:
                fieldPath: spec.nodeName
```

Of course the downside would be that we will have to maintain these images as well.